File: teacher-webapp/src/utils/commandResultFormatter.js
import { ROUTES } from "../constants/routes";

// Routes the app actually serves (see constants/routes.js). A NAVIGATE step whose
// target is outside these prefixes is treated as a planner hallucination and
// dropped rather than navigated to.
const KNOWN_ROUTE_PREFIXES = ["/classrooms", "/content/"];

// The platform API serializes Mongo's `_id` as `id` (see platform/app/models/*).
// Older backend-server responses used `_id`. Read both so a serializer change can
// never silently break navigation again — a missing id used to make the result
// card render with no action at all.
const idOf = (obj) => obj?.id || obj?._id || null;

// Endpoints differ: /class returns a plain array, /content returns
// { data: [...], pagination }. Accept either.
const unwrapList = (data) =>
  Array.isArray(data) ? data : Array.isArray(data?.data) ? data.data : [];

const personName = (p) => p?.name || p?.phone_number || p?.phoneNumber || "Unnamed";

/**
 * Format a single command+result pair into a display-friendly object.
 * @param {Object} command - { method, path, description, ... }
 * @param {Object} result  - { status, data, error }
 * @returns {{ title: string, summary: string, items: string[] }}
 */
export function formatResult(command, result) {
  if (!command || !result) {
    return { title: "Command", summary: "No details available", items: [] };
  }

  if (result.error) {
    return { title: command.description || "Command", summary: result.error, items: [] };
  }

  const path = command.path || "";
  const data = result.data;

  // Single classroom (GET /class/:id) — list its members instead of a bare
  // "Completed successfully", which told the teacher nothing.
  if (path.match(/^\/class\/[^/]+$/) && command.method === "GET") {
    const students = Array.isArray(data?.students) ? data.students : [];
    const leaders = Array.isArray(data?.leaders) ? data.leaders : [];
    return {
      title: data?.name ? `Class: ${data.name}` : "Classroom",
      summary:
        `${students.length} student${students.length !== 1 ? "s" : ""}, ` +
        `${leaders.length} leader${leaders.length !== 1 ? "s" : ""}`,
      items: [...students, ...leaders].map(personName),
    };
  }

  // Classroom list
  if (path.match(/\/class\/?$/) && command.method === "GET") {
    const names = unwrapList(data).map((c) => c.name);
    return {
      title: "Classrooms",
      summary: `Found ${names.length} classroom${names.length !== 1 ? "s" : ""}`,
      items: names,
    };
  }

  // Students list
  if (path.includes("/teacher/students")) {
    const names = unwrapList(data).map(personName);
    return {
      title: "Students",
      summary: `Found ${names.length} student${names.length !== 1 ? "s" : ""}`,
      items: names,
    };
  }

  // Teacher profile
  if (path.includes("/teacher/me")) {
    const phone = data?.phoneNumber || data?.phone || "";
    return {
      title: "Your Profile",
      summary: phone ? `Phone: ${phone}` : "Profile loaded",
      items: data?.name ? [data.name] : [],
    };
  }

  // Content list (handles both plain array and paginated { data: [...], pagination })
  if (path.match(/\/content\/?/) && command.method === "GET") {
    const items = Array.isArray(data) ? data : Array.isArray(data?.data) ? data.data : [];
    const titles = items.map((c) => c.title?.english || c.title?.local || c.expName || c.name || "Untitled");
    return {
      title: "Content",
      summary: `Found ${titles.length} item${titles.length !== 1 ? "s" : ""}`,
      items: titles,
    };
  }

  // Fallback
  return {
    title: command.description || "Command",
    summary: result.status < 300 ? "Completed successfully" : `Status ${result.status}`,
    items: [],
  };
}

/**
 * Determine which page to navigate to based on the executed commands.
 * For content commands, extracts the content ID to enable direct navigation + auto-play.
 * @param {Object[]} commands - Array of command objects
 * @param {Object[]} results  - Array of result objects
 * @returns {{ label: string, path: string, autoNavigate?: boolean } | null}
 */
export function getNavigationTarget(commands, results) {
  if (!commands || commands.length === 0) return null;

  let confIdSearchResult = null;
  let classIdSearchResult = null;
  let singleClassData = null;
  let sawClassCommand = false;
  let sawStudentsCommand = false;

  for (let i = 0; i < commands.length; i++) {
    const cmd = commands[i];
    const res = results?.[i];
    const path = cmd.path || "";

    // Frontend navigation pseudo-command — go straight to the requested route.
    if (cmd.method === "NAVIGATE" && res?.status < 300) {
      const target = res?.data?.navigate || path;
      if (target === "/action/content-load-more") {
        return { action: "CONTENT_LOAD_MORE", label: "Load more content", autoNavigate: true };
      }
      if (target === "/action/content-stop") {
        return { action: "CONTENT_STOP", label: "Stop content", autoNavigate: true };
      }
      // The content library is a drawer, not a route. The planner sometimes invents
      // a page for it ("/content-library", "/content"); map those onto the drawer
      // instead of navigating to a route that does not exist (blank screen).
      if (/^\/(action\/)?content(-library|s)?\/?$/.test(target)) {
        return { action: "OPEN_CONTENT_DRAWER", label: "Open Content Library", autoNavigate: true };
      }
      // Only navigate to routes the app actually serves — an unknown invented path
      // would otherwise render nothing at all.
      if (!KNOWN_ROUTE_PREFIXES.some((p) => target.startsWith(p))) {
        return null;
      }
      return { label: "Go", path: target, autoNavigate: true };
    }

    // Track classroom ID if fetched
    if (path.match(/^\/class\/([^/]+)$/) && cmd.method === "GET" && res?.status < 300) {
      classIdSearchResult = idOf(res?.data);
      singleClassData = res?.data;
    }

    // Track conference ID if created
    if (path.match(/\/conference\/create/) && res?.status < 300) {
      confIdSearchResult = res?.data?.id;
    }

    // Conference started -> Navigate to Classroom Detail with autoStart state
    if (path.match(/\/conference\/start/) && res?.status < 300) {
      const targetConfId = confIdSearchResult || path.split("/").pop(); // Fallback to path param
      if (classIdSearchResult) {
         return {
           label: "Go to Conference Call",
           path: ROUTES.CLASSROOM_DETAIL(classIdSearchResult),
           autoNavigate: true,
           state: { confId: targetConfId, autoStart: true },
         };
      }
    }

    // Content command — navigate directly to the content detail page for auto-play
    if (path.match(/\/content/) && cmd.method === "GET" && res?.status < 300) {
      const raw = res?.data;
      // Unwrap paginated response { data: [...], pagination } or plain array
      const items = Array.isArray(raw) ? raw : Array.isArray(raw?.data) ? raw.data : null;
      const single = !items && idOf(raw) ? raw : null;

      // Single content item → play directly
      if (single) {
        return {
          label: `Play: ${single.title?.english || single.expName || "Content"}`,
          path: ROUTES.CONTENT_DETAILS(idOf(single)),
          autoNavigate: true,
        };
      }

      // Array result
      if (items && items.length > 0 && idOf(items[0])) {
        // Search query (search/expName/ids) → play the first match directly.
        // `search=` is the title search, which is what "play The alphabet song" uses.
        if (path.includes("search=") || path.includes("expName=") || path.includes("ids=")) {
          return {
            label: `Play: ${items[0].title?.english || items[0].expName || "Content"}`,
            path: ROUTES.CONTENT_DETAILS(idOf(items[0])),
            autoNavigate: true,
            state: { contentList: items, currentIndex: 0 },
          };
        }
        // Content library browse (no search filter) → open content drawer
        return { action: "OPEN_CONTENT_DRAWER", label: "Open Content Library", autoNavigate: true };
      }

      // No items resolved
      return null;
    }

    // New classroom created -> offer to navigate directly to it (POST with no _id = create)
    if (path.match(/\/class\/?$/) && cmd.method === "POST" && res?.status < 300 && idOf(res?.data)) {
      const roomName = res.data.name || "new classroom";
      return {
        label: `Go to ${roomName}`,
        path: ROUTES.CLASSROOM_DETAIL(idOf(res.data)),
      };
    }

    // Note generic class/student commands but DON'T return yet — a later command
    // (e.g. conference start) may be a higher-priority navigation target.
    if (path.match(/\/class/)) sawClassCommand = true;
    if (path.includes("/teacher/students")) sawStudentsCommand = true;
  }

  // Generic fallback: only after scanning ALL commands for priority targets.
  // A standalone "open class X" (GET /class/:id, no conference-start after it) goes
  // straight to that class rather than the generic classrooms list.
  if (idOf(singleClassData)) {
    return { label: `Go to ${singleClassData.name || "classroom"}`, path: ROUTES.CLASSROOM_DETAIL(idOf(singleClassData)), autoNavigate: true };
  }
  if (sawClassCommand || sawStudentsCommand) {
    return { label: "Go to Classrooms", path: ROUTES.CLASSROOMS };
  }

  return null;
}

File: Teacher-App/app/src/main/java/com/example/seeds/ui/voiceCommand/CommandResultFormatter.kt
package com.example.seeds.ui.voiceCommand

import android.os.Parcelable
import androidx.navigation.NavDirections
import com.example.seeds.R
import com.example.seeds.model.AudioContent
import com.example.seeds.model.Classroom
import com.example.seeds.model.Content
import com.example.seeds.model.LocalizedContent
import com.example.seeds.model.Student
import com.example.seeds.network.CommandResult
import com.example.seeds.network.VoiceCommand
import com.example.seeds.ui.call.CallSettingsFragmentDirections

// Android port of teacher-webapp/src/utils/commandResultFormatter.js — kept 1:1 with web
// semantics (same path matches, same priority order) so both clients show the same thing.
// Pure functions: the parser logic lives here and is unit-tested (per plan Phase 5).

data class FormattedResult(
    val title: String,
    val summary: String,
    val items: List<String>,
    val isSuccess: Boolean
)

data class NavigationTarget(
    val label: String,
    // R.id.* nav destination. null = no reachable destination; button renders label-only.
    val destinationId: Int? = null,
    val autoNavigate: Boolean = false,
    // Parcelable nav-arg for destinations that require one (Content for the player, Classroom
    // for call settings), plus its safe-args bundle key. Built from the backend result map by
    // buildContent/buildClassroom below. null = argless destination.
    val navArg: Parcelable? = null,
    val navArgKey: String? = null,
    // Pre-built multi-arg navigation (e.g. call_nav needs BOTH phoneNumbers and classroom —
    // more than the single navArg above can carry). Takes priority over destinationId/navArg
    // when set.
    val directions: NavDirections? = null
)

private fun CommandResult?.isOk(): Boolean = this != null && error == null && status in 0 until 300

@Suppress("UNCHECKED_CAST")
private fun Any?.asList(): List<Map<*, *>> = when (this) {
    is List<*> -> this.filterIsInstance<Map<*, *>>()
    is Map<*, *> -> (this["data"] as? List<*>)?.filterIsInstance<Map<*, *>>() ?: emptyList()
    else -> emptyList()
}

private fun Map<*, *>.str(key: String): String? = this[key]?.toString()

fun formatResult(command: VoiceCommand?, result: CommandResult?): FormattedResult {
    val isSuccess = result.isOk()
    if (command == null || result == null) {
        return FormattedResult("Command", "No details available", emptyList(), false)
    }
    if (result.error != null) {
        return FormattedResult(command.description ?: "Command", result.error, emptyList(), false)
    }

    val path = command.path
    val data = result.data

    // Classroom list
    if (Regex("""/class/?$""").containsMatchIn(path) && command.method == "GET") {
        val names = data.asList().mapNotNull { it.str("name") }
        return FormattedResult("Classrooms", "Found ${names.size} classroom${if (names.size != 1) "s" else ""}", names, isSuccess)
    }

    // Students list
    if (path.contains("/teacher/students")) {
        val names = data.asList().mapNotNull { it.str("name") ?: it.str("phoneNumber") }
        return FormattedResult("Students", "Found ${names.size} student${if (names.size != 1) "s" else ""}", names, isSuccess)
    }

    // Teacher profile
    if (path.contains("/teacher/me")) {
        val map = data as? Map<*, *>
        val phone = map?.str("phoneNumber") ?: map?.str("phone")
        return FormattedResult(
            "Your Profile",
            if (!phone.isNullOrEmpty()) "Phone: $phone" else "Profile loaded",
            map?.str("name")?.let { listOf(it) } ?: emptyList(),
            isSuccess
        )
    }

    // Content list
    if (path.contains("/content") && command.method == "GET") {
        val titles = data.asList().map { item ->
            val title = item["title"] as? Map<*, *>
            title?.str("english") ?: title?.str("local") ?: item.str("expName") ?: item.str("name") ?: "Untitled"
        }
        return FormattedResult("Content", "Found ${titles.size} item${if (titles.size != 1) "s" else ""}", titles, isSuccess)
    }

    // Fallback
    return FormattedResult(
        command.description ?: "Command",
        if (result.status < 300) "Completed successfully" else "Status ${result.status}",
        emptyList(),
        isSuccess
    )
}

/**
 * Which screen to offer navigation to, from the executed commands. Mirrors the web
 * priority order: NAVIGATE > conference-start > content > new-classroom > generic class/student.
 * destinationId is set only for targets reachable without a typed nav-arg; the rest carry a
 * label (button renders) and are wired to arg-passing navigation in Phase 8.
 */
fun getNavigationTarget(commands: List<VoiceCommand>, results: List<CommandResult>): NavigationTarget? {
    if (commands.isEmpty()) return null

    var classId: String? = null
    var singleClassData: Map<*, *>? = null
    var confId: String? = null
    var confCreateBody: Map<*, *>? = null
    var sawClassCommand = false
    var sawStudentsCommand = false

    for (i in commands.indices) {
        val cmd = commands[i]
        val res = results.getOrNull(i)
        val path = cmd.path

        // Frontend navigation pseudo-command. Only "go home"/"go to classrooms" resolves to a
        // real Android destination today — other NAVIGATE intents (content-load-more/stop) have
        // no Android UI surface yet, so they fall through to the label-only no-op below.
        if (cmd.method == "NAVIGATE" && res.isOk()) {
            val target = (res?.data as? Map<*, *>)?.str("navigate") ?: path
            if (target.contains("/classrooms") && !target.contains("/detail/")) {
                return NavigationTarget("Go to Classrooms", destinationId = R.id.classroomFragment, autoNavigate = true)
            }
            return NavigationTarget("Go", autoNavigate = true)
        }

        if (Regex("""^/class/([^/]+)$""").containsMatchIn(path) && cmd.method == "GET" && res.isOk()) {
            val map = res?.data as? Map<*, *>
            classId = map?.str("_id")
            singleClassData = map
        }

        if (path.contains("/conference/create") && res.isOk()) {
            confId = (res?.data as? Map<*, *>)?.str("id")
            confCreateBody = cmd.body as? Map<*, *>
        }

        // Conference started -> go straight into the call (mirrors CallSettingsFragment's own
        // "Start Conference" construction: phoneNumbers from the create request, classroom from
        // the earlier GET /class/:id fetch in this same command batch).
        if (path.contains("/conference/start") && res.isOk() && classId != null) {
            val classroom = singleClassData?.let { buildClassroom(it) }
            val phones = (confCreateBody?.get("student_phones") as? List<*>)?.mapNotNull { it?.toString() }
            if (classroom != null && !phones.isNullOrEmpty()) {
                val leader = confCreateBody?.str("leader_phone")
                val directions = CallSettingsFragmentDirections
                    .actionCallSettingsFragmentToCallNav(phones.toTypedArray(), classroom)
                    .setLeader(leader)
                return NavigationTarget("Go to Conference Call", autoNavigate = true, directions = directions)
            }
            // Missing phones/classroom data — fall back to the classroom rather than a dead nav.
            return NavigationTarget("Go to Classrooms", destinationId = R.id.classroomFragment, autoNavigate = true)
        }

        // Content command -> deep-link to the player (ContentDetailsFragment autoplays from the arg)
        if (path.contains("/content") && cmd.method == "GET" && res.isOk()) {
            val single = (res?.data as? Map<*, *>)?.takeIf { it["_id"] != null }
            if (single != null) return contentTarget(single)
            val items = res?.data.asList()
            if (items.isNotEmpty() && items[0]["_id"] != null) {
                if (path.contains("expName=") || path.contains("ids=")) return contentTarget(items[0])
                // No filter = "browse the library" -> the Content tab (argless-navigable).
                return NavigationTarget("Open Content Library", destinationId = R.id.homeFragment)
            }
            return null
        }

        // New classroom created -> open its call-settings screen (Android's per-class screen)
        if (Regex("""/class/?$""").containsMatchIn(path) && cmd.method == "POST" && res.isOk()) {
            val map = res?.data as? Map<*, *>
            if (map?.str("_id") != null) {
                return classTarget("Go to ${map.str("name") ?: "new classroom"}", map)
            }
        }

        if (path.contains("/class")) sawClassCommand = true
        if (path.contains("/teacher/students")) sawStudentsCommand = true
    }

    // A standalone "open class X" (GET /class/:id, no conference-start after it) goes
    // straight to that class rather than the generic classrooms list.
    singleClassData?.let { map ->
        if (map.str("_id") != null) {
            val target = classTarget("Go to ${map.str("name") ?: "classroom"}", map)
            // Only auto-jump when a real per-class target was built, not the safe list fallback.
            return if (target.destinationId == R.id.callSettingsFragment) target.copy(autoNavigate = true) else target
        }
    }

    if (sawClassCommand || sawStudentsCommand) {
        return NavigationTarget("Go to Classrooms", destinationId = R.id.classroomFragment)
    }
    return null
}

// --- nav-arg construction: raw backend result map -> typed Parcelable model ---------------
// Built by hand (not Moshi): the app's Moshi is Kotshi-codegen-only and Content has no adapter,
// and this only needs the handful of fields the player/call-settings screens actually read.

private fun mapOfLocalized(value: Any?): LocalizedContent? {
    val m = value as? Map<*, *> ?: return null
    val english = m.str("english") ?: return null
    return LocalizedContent(english = english, local = m.str("local"), audioUrl = m.str("audioUrl"))
}

fun buildContent(map: Map<*, *>): Content? {
    val id = map.str("_id") ?: return null
    val audio = (map["audioContent"] as? List<*>).orEmpty().mapNotNull { entry ->
        val am = entry as? Map<*, *> ?: return@mapNotNull null
        val url = am.str("audioUrl") ?: return@mapNotNull null
        AudioContent(description = am.str("description") ?: "", audioUrl = url)
    }
    return Content(
        _id = id,
        type = map.str("type") ?: "",
        description = map.str("description"),
        language = map.str("language") ?: "en",
        title = mapOfLocalized(map["title"]) ?: LocalizedContent(),
        theme = mapOfLocalized(map["theme"]) ?: LocalizedContent(),
        audioContent = audio
    )
}

fun buildClassroom(map: Map<*, *>): Classroom? {
    val name = map.str("name") ?: return null
    fun students(key: String): List<Student> =
        (map[key] as? List<*>).orEmpty().mapNotNull { entry ->
            val sm = entry as? Map<*, *> ?: return@mapNotNull null
            val phone = sm.str("phoneNumber") ?: sm.str("phone") ?: return@mapNotNull null
            Student(
                phoneNumber = phone,
                name = sm.str("name") ?: "",
                isLeader = sm["isLeader"] as? Boolean ?: false,
                _id = sm.str("_id")
            )
        }
    return Classroom(
        _id = map.str("_id"),
        name = name,
        teacher = map.str("teacher") ?: "",
        students = students("students"),
        leaders = students("leaders"),
        contentIds = (map["contentIds"] as? List<*>).orEmpty().mapNotNull { it?.toString() }
    )
}

private fun contentTarget(map: Map<*, *>): NavigationTarget {
    val content = buildContent(map)
    val title = content?.title?.english ?: map.str("expName") ?: "Content"
    // Auto-navigate into the player so "play X" actually starts playing (web parity).
    return if (content != null) {
        NavigationTarget(
            "Play: $title", R.id.contentDetailsFragment,
            autoNavigate = true, navArg = content, navArgKey = "content"
        )
    } else {
        NavigationTarget("Play: $title")
    }
}

private fun classTarget(label: String, map: Map<*, *>): NavigationTarget {
    val classroom = buildClassroom(map)
    // No usable class model -> fall back to the safe classroom list rather than a dead/crashing nav.
    return if (classroom != null) {
        NavigationTarget(label, R.id.callSettingsFragment, navArg = classroom, navArgKey = "classroom")
    } else {
        NavigationTarget("Go to Classrooms", destinationId = R.id.classroomFragment)
    }
}
